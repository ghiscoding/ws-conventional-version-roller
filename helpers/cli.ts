import execa from 'execa';
import path from 'path';

import lernaCLI from '../packages/cli/src/lerna-cli';

const LERNA_BIN = path.resolve(__dirname, '../packages/cli/src/cli.ts');

/**
 * A higher-order function to help with passing _actual_ yargs-parsed argv
 * into command constructors (instead of artificial direct parameters).
 *
 * @param {Object} commandModule The yargs command exports
 * @return {Function} with partially-applied yargs config
 */
export function commandRunner(commandModule: any) {
  /* eslint-disable import/no-dynamic-require, global-require */
  const cmd = commandModule.command.split(' ')[0];

  // prime the pump so slow-as-molasses CI doesn't fail with delayed require()
  require(path.resolve((require as any).main.filename, '../..'));

  return (cwd: string) => {
    // create a _new_ yargs instance every time cwd changes to avoid singleton pollution
    const cli = lernaCLI([], cwd)
      .exitProcess(false)
      .detectLocale(false)
      .showHelpOnFail(false)
      .wrap(null)
      .command(commandModule);

    return (...args: string[]) =>
      new Promise((resolve, reject) => {
        const yargsMeta: any = {};
        const context = {
          cwd,
          lernaVersion: '__TEST_VERSION__',
          loglevel: 'silent',
          progress: false,
          onResolved: (result) => {
            // success resolves the result, if any, returned from execute()
            resolve(Object.assign({}, result, yargsMeta));
          },
          onRejected: (result) => {
            Object.assign(result, yargsMeta);
            // tests expect errors thrown to indicate failure,
            // _not_ just non-zero exitCode
            reject(result);
          },
        };

        const parseFn = (_yargsError: Error | undefined, parsedArgv: any, yargsOutput: string) => {
          // this is synchronous, before the async handlers resolve
          Object.assign(yargsMeta, { parsedArgv, yargsOutput });
        };

        cli
          .fail((msg: string, err: Error) => {
            // since yargs 10.1.0, this is the only way to catch handler rejection
            // _and_ yargs validation exceptions when using async command handlers
            const actual = (err || new Error(msg)) as Error & { exitCode: number };
            // backfill exitCode for test convenience
            yargsMeta.exitCode = 'exitCode' in actual ? actual.exitCode : 1;
            context.onRejected(actual);
          })
          .parse([cmd, ...args], context, parseFn);
      });
  };
}

exports.cliRunner = function cliRunner(cwd: string, env: { [key: string]: string }) {
  const opts = {
    cwd,
    env: Object.assign(
      {
        CI: 'true',
        // always turn off chalk
        FORCE_COLOR: '0',
      },
      env
    ),
    // when debugging integration test snapshots, uncomment next line
    // stdio: ["ignore", "inherit", "inherit"],
  };

  return (...args: string[]) => execa('node', [LERNA_BIN].concat(args), opts);
};
