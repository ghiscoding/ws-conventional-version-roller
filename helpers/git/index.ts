import execa from 'execa';
import os from 'os';
import { tempWrite } from '../../packages/version/src/utils/temp-write';
import path from 'path';
import cp from 'child_process';
import loadJsonFile from 'load-json-file';
import writeJsonFile from 'write-json-file';
import gitSHA from '../serializers/serialize-git-sha';

// Contains all relevant git config (user, commit.gpgSign, etc)
const TEMPLATE = path.resolve(__dirname, 'template');

export function getCommitMessage(cwd, format = '%B') {
  return execa('git', ['log', '-1', `--pretty=format:${format}`], { cwd }).then((result) => result.stdout);
}

export function gitAdd(cwd, ...files) {
  return execa('git', ['add', ...files], { cwd });
}

export function gitCheckout(cwd, args) {
  return execa('git', ['checkout', ...args], { cwd });
}

export function gitCommit(cwd, message) {
  if (message.indexOf(os.EOL) > -1) {
    // Use tempfile to allow multi\nline strings.
    return tempWrite(message).then((fp) => execa('git', ['commit', '-F', fp], { cwd }));
  }

  return execa('git', ['commit', '-m', message], { cwd });
}

export function gitInit(cwd, ...args) {
  return execa('git', ['init', '--template', TEMPLATE, ...args], { cwd }).then(() =>
    execa('git', ['checkout', '-B', 'main'], { cwd })
  );
}

export function gitMerge(cwd, args) {
  return execa('git', ['merge', ...args], { cwd });
}

export function gitStatus(cwd) {
  return cp.spawnSync('git', ['status', '--porcelain'], { cwd, encoding: 'utf8' });
}

export function gitTag(cwd, tagName) {
  return execa('git', ['tag', tagName, '-m', tagName], { cwd });
}

export function showCommit(cwd, ...args) {
  return execa(
    'git',
    [
      'show',
      '--unified=0',
      '--ignore-space-at-eol',
      '--pretty=%B%+D',
      // make absolutely certain that no OS localization
      // changes the expected value of the path prefixes
      '--src-prefix=a/',
      '--dst-prefix=b/',
      ...args,
    ],
    { cwd }
  ).then((result) => gitSHA.serialize(result.stdout));
}

export function commitChangeToPackage(cwd, packageName, commitMsg, data) {
  const packageJSONPath = path.join(cwd, 'packages', packageName, 'package.json');

  // QQ no async/await yet...
  let chain: Promise<any> = Promise.resolve();

  chain = chain.then(() => loadJsonFile(packageJSONPath));
  chain = chain.then((pkg) => writeJsonFile(packageJSONPath, Object.assign(pkg, data)));
  chain = chain.then(() => gitAdd(cwd, packageJSONPath));
  chain = chain.then(() => gitCommit(cwd, commitMsg));

  return chain;
}
