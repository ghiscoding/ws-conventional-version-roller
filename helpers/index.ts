import normalizeNewline from 'normalize-newline';
import normalizePath from 'normalize-path';
import { relative } from 'node:path';

import { Project } from '../packages/core/src/project/index.js';

/**
 * Update lerna config inside a test case.
 *
 * @param {String} testDir where target lerna.json exists
 * @param {Object} updates mixed into existing JSON via Object.assign
 */
export function updateLernaConfig(testDir, updates) {
  const project = new Project(testDir);

  Object.assign(project.config, updates);

  return project.serializeConfig();
}

export function multiLineTrimRight(str) {
  return normalizeNewline(str)
    .split('\n')
    .map((line) => line.trimRight())
    .join('\n');
}

export function normalizeRelativeDir(testDir, filePath) {
  return normalizePath(relative(testDir, filePath));
}

export * from './git/index.js';
export * from './fixtures.js';
export * from './npm/index.js';
export * from './cli.js';
export * from './logging-output.js';
export * from './pkg-matchers.js';
