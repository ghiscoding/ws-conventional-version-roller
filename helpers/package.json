{
  "name": "@lerna-test/helpers",
  "version": "0.0.0-test-only",
  "description": "Private test helpers for the lerna-lite monorepo",
  "main": "index.js",
  "private": true,
  "license": "MIT"
}
