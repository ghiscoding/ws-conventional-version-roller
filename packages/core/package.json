{
  "name": "@lerna-lite/core",
  "description": "Lerna-Lite core implementation module",
  "version": "2.1.0",
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "/dist"
  ],
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "module": "dist/index.js",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "default": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./*": "./*",
    "./cli.js": "./dist/cli.js",
    "./dist/cli.js": "./dist/cli.js"
  },
  "typesVersions": {
    "*": {
      "*": [
        "./dist/index.d.ts"
      ]
    }
  },
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "tsc",
    "build:incremental": "tsc --incremental --sourceMap false --excludeDirectories dist",
    "pack-tarball": "npm pack"
  },
  "license": "MIT",
  "author": "Ghislain B.",
  "homepage": "https://github.com/lerna-lite/lerna-lite",
  "repository": {
    "type": "git",
    "url": "https://github.com/lerna-lite/lerna-lite.git",
    "directory": "packages/core"
  },
  "bugs": {
    "url": "https://github.com/lerna-lite/lerna-lite/issues"
  },
  "engines": {
    "node": ">=16.15.0",
    "npm": ">=8.5.0"
  },
  "dependencies": {
    "@npmcli/run-script": "^6.0.0",
    "chalk": "^5.2.0",
    "clone-deep": "^4.0.1",
    "config-chain": "^1.1.13",
    "cosmiconfig": "^8.1.3",
    "dedent": "^0.7.0",
    "execa": "^7.1.1",
    "fs-extra": "^11.1.1",
    "glob-parent": "^6.0.2",
    "globby": "^13.1.4",
    "inquirer": "^9.1.5",
    "is-ci": "^3.0.1",
    "json5": "^2.2.3",
    "load-json-file": "^7.0.1",
    "minimatch": "^9.0.0",
    "npm-package-arg": "^10.1.0",
    "npmlog": "^7.0.1",
    "p-map": "^6.0.0",
    "p-queue": "^7.3.4",
    "resolve-from": "^5.0.0",
    "semver": "^7.5.0",
    "slash": "^5.0.0",
    "strong-log-transformer": "^2.1.0",
    "write-file-atomic": "^5.0.0",
    "write-json-file": "^5.0.0",
    "write-pkg": "^5.1.0"
  },
  "devDependencies": {
    "@types/clone-deep": "^4.0.1",
    "@types/dedent": "^0.7.0",
    "@types/fs-extra": "^11.0.1",
    "@types/glob-parent": "^5.1.1",
    "@types/inquirer": "^9.0.3",
    "@types/minimatch": "^5.1.2",
    "@types/npm-package-arg": "^6.1.1",
    "@types/semver": "^7.3.13",
    "@types/strong-log-transformer": "^1.0.0",
    "@types/write-file-atomic": "^4.0.0",
    "npm-registry-fetch": "^14.0.4",
    "yargs": "^17.7.1",
    "yargs-parser": "^21.1.1"
  },
  "funding": {
    "type": "ko_fi",
    "url": "https://ko-fi.com/ghiscoding"
  }
}
