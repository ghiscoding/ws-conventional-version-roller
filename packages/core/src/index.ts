// folders
export * from './models';
export * from './package-graph';
export * from './project';
export * from './utils';

// files
export * from './child-process';
export * from './constants';
export * from './command';
export * from './package';
export * from './prompt';
export * from './validation-error';
