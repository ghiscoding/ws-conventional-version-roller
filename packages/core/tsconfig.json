{
  "extends": "../../tsconfig.base.json",
  "compileOnSave": false,
  "compilerOptions": {
    "module": "ESNext",
    "rootDir": "src",
    "outDir": "dist",
    "types": ["node"]
  },
  "filesGlob": ["./src/**/*.ts"],
  "include": ["src/**/*.ts"],
  "exclude": ["dist", "node_modules", "**/__helpers__/**", "**/__mocks__/**", "**/__tests__/**"]
}
