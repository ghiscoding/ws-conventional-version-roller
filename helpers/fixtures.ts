import execa from 'execa';
import fileUrl from 'file-url';
import tempy from 'tempy';
import fs from 'fs-extra';
import findUp from 'find-up';
import path from 'path';
import { gitAdd, gitCommit, gitInit } from './git';

export function cloneFixtureFactory(startDir: string) {
  const initFixture = initFixtureFactory(startDir);

  return (...args: any[]) =>
    // @ts-ignore
    initFixture(...args).then((cwd) => {
      const repoDir = tempy.directory();
      const repoUrl = fileUrl(repoDir, { resolve: false });

      return execa('git', ['init', '--bare'], { cwd: repoDir })
        .then(() => execa('git', ['checkout', '-B', 'main'], { cwd }))
        .then(() => execa('git', ['remote', 'add', 'origin', repoUrl], { cwd }))
        .then(() => execa('git', ['push', '-u', 'origin', 'main'], { cwd }))
        .then(() => ({
          cwd,
          repository: repoUrl,
        }));
    });
}

export function findFixture(cwd: string, fixtureName: string) {
  return findUp(path.join('__fixtures__', fixtureName), { cwd, type: 'directory' }).then((fixturePath) => {
    if (fixturePath === undefined) {
      throw new Error(`Could not find fixture with name "${fixtureName}"`);
    }

    return fixturePath;
  });
}

export function copyFixture(targetDir: string, fixtureName: string, cwd: string) {
  return findFixture(cwd, fixtureName).then((fp) => fs.copy(fp, targetDir));
}

export function initFixtureFactory(startDir: string) {
  return (fixtureName: string, commitMessage: boolean | string = 'Init commit') => {
    const cwd = tempy.directory();
    let chain: Promise<any> = Promise.resolve();

    chain = chain.then(() => process.chdir(cwd));
    chain = chain.then(() => copyFixture(cwd, fixtureName, startDir));
    chain = chain.then(() => gitInit(cwd, '.'));

    if (commitMessage) {
      chain = chain.then(() => gitAdd(cwd, '-A'));
      chain = chain.then(() => gitCommit(cwd, commitMessage));
    }

    return chain.then(() => cwd);
  };
}

export function initNamedFixtureFactory(startDir: string) {
  return (dirName: string, fixtureName: string, commitMessage: boolean | string = 'Init commit') => {
    const cwd = path.join(tempy.directory(), dirName);
    let chain: Promise<any> = Promise.resolve();

    chain = chain.then(() => fs.ensureDir(cwd));
    chain = chain.then(() => process.chdir(cwd));
    chain = chain.then(() => copyFixture(cwd, fixtureName, startDir));
    chain = chain.then(() => gitInit(cwd, '.'));

    if (commitMessage) {
      chain = chain.then(() => gitAdd(cwd, '-A'));
      chain = chain.then(() => gitCommit(cwd, commitMessage));
    }

    return chain.then(() => cwd);
  };
}
