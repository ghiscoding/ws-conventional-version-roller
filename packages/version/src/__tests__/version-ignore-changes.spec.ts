import nodeFs from 'fs';
jest.spyOn(nodeFs, 'renameSync');

// local modules _must_ be explicitly mocked
jest.mock('../lib/git-push', () => jest.requireActual('../lib/__mocks__/git-push'));
jest.mock('../lib/is-anything-committed', () => jest.requireActual('../lib/__mocks__/is-anything-committed'));
jest.mock('../lib/is-behind-upstream', () => jest.requireActual('../lib/__mocks__/is-behind-upstream'));
jest.mock('../lib/remote-branch-exists', () => jest.requireActual('../lib/__mocks__/remote-branch-exists'));

// mocked modules of @lerna-lite/core
jest.mock('@lerna-lite/core', () => ({
  ...(jest.requireActual('@lerna-lite/core') as any), // return the other real methods, below we'll mock only 2 of the methods
  logOutput: jest.requireActual('../../../core/src/__mocks__/output').logOutput,
  promptConfirmation: jest.requireActual('../../../core/src/__mocks__/prompt').promptConfirmation,
  promptSelectOne: jest.requireActual('../../../core/src/__mocks__/prompt').promptSelectOne,
  promptTextInput: jest.requireActual('../../../core/src/__mocks__/prompt').promptTextInput,
  throwIfUncommitted: jest.requireActual('../../../core/src/__mocks__/check-working-tree').throwIfUncommitted,
}));

import fs from 'fs-extra';
import path from 'path';
import yargParser from 'yargs-parser';

// helpers
import { gitAdd } from '@lerna-test/helpers';
import { gitCommit } from '@lerna-test/helpers';
import { gitTag } from '@lerna-test/helpers';
import { showCommit } from '@lerna-test/helpers';
import { initFixtureFactory } from '@lerna-test/helpers';
const initFixture = initFixtureFactory(path.resolve(__dirname, '../../../publish/src/__tests__'));

// test command
import { VersionCommand } from '../version-command';

// stabilize commit SHA
import gitSHA from '@lerna-test/helpers/serializers/serialize-git-sha';
import { VersionCommandOption } from '@lerna-lite/core';
expect.addSnapshotSerializer(gitSHA);

const createArgv = (cwd, ...args) => {
  args.unshift('version');
  if (args.length > 0 && args[1]?.length > 0 && !args[1].startsWith('-')) {
    args[1] = `--bump=${args[1]}`;
  }
  const parserArgs = args.map(String);
  const argv = yargParser(parserArgs, { array: [{ key: 'ignoreChanges' }] });
  argv['$0'] = cwd;
  return argv as unknown as VersionCommandOption;
};

describe('version --ignore-changes', () => {
  const setupChanges = async (cwd, tuples) => {
    await gitTag(cwd, 'v1.0.0');
    await Promise.all(tuples.map(([filePath, content]) => fs.outputFile(path.join(cwd, filePath), content, 'utf8')));
    await gitAdd(cwd, '.');
    await gitCommit(cwd, 'setup');
  };

  it('does not version packages with ignored changes', async () => {
    const cwd = await initFixture('normal');

    await setupChanges(cwd, [
      ['packages/package-2/README.md', 'oh'],
      ['packages/package-3/__tests__/pkg3.test.js', 'hai'],
      ['packages/package-4/lib/foo.js', 'there'],
    ]);

    // await lernaVersion(cwd)(
    await new VersionCommand(
      createArgv(
        cwd,
        '--ignore-changes',
        'README.md',

        '--ignore-changes',
        '**/__tests__/**',

        '--ignore-changes',
        'package-4' // notably does NOT work, needs to be "**/package-4/**" to match
      )
    );

    const changedFiles = await showCommit(cwd, '--name-only');
    expect(changedFiles).toMatchSnapshot();
  });
});
