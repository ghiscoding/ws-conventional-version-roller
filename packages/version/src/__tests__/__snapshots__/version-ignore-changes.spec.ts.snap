// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`version --ignore-changes > does not version packages with ignored changes 1`] = `
v1.0.1

HEAD -> main, tag: v1.0.1

lerna.json
packages/package-4/package.json
`;
